src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/FindPal.cs

```diff
@@ -0,0 +1,405 @@
+                                    CERT_NAME_VALUE* pNameValue = (CERT_NAME_VALUE*)pvDecoded;
+                                    string actual = Marshal.PtrToStringUni(new IntPtr(pNameValue->Value.pbData));
+                                    if (templateName.Equals(actual, StringComparison.OrdinalIgnoreCase))
+                                        foundMatch = true;
+                                }))
+                            {
+                                return false;
+                            }
+                        }
+                    }
+
+                    if (!foundMatch)
+                    {
+                        CERT_EXTENSION* pV2Template = Interop.crypt32.CertFindExtension(Oids.CertificateTemplate,
+                            pCertInfo->cExtension, pCertInfo->rgExtension);
+                        if (pV2Template != null)
+                        {
+                            byte[] extensionRawData = pV2Template->Value.ToByteArray();
+                            if (!extensionRawData.DecodeObjectNoThrow(
+                                CryptDecodeObjectStructType.X509_CERTIFICATE_TEMPLATE,
+                                delegate(void* pvDecoded)
+                                {
+                                    CERT_TEMPLATE_EXT* pTemplateExt = (CERT_TEMPLATE_EXT*)pvDecoded;
+                                    string actual = Marshal.PtrToStringAnsi(pTemplateExt->pszObjId);
+                                    string expectedOidValue =
+                                        OidInfo.FindOidInfo(CryptOidInfoKeyType.CRYPT_OID_INFO_NAME_KEY, templateName,
+                                            OidGroup.Template, fallBackToAllGroups: true).OID;
+                                    if (expectedOidValue == null)
+                                        expectedOidValue = templateName;
+                                    if (expectedOidValue.Equals(actual, StringComparison.OrdinalIgnoreCase))
+                                        foundMatch = true;
+                                }))
+                            {
+                                return false;
+                            }
+                        }
+                    }
+
+                    GC.KeepAlive(pCertContext);
+                    return foundMatch;
+                });
+        }
+
+        public unsafe void FindByApplicationPolicy(string oidValue)
+        {
+            FindCore(
+                delegate (SafeCertContextHandle pCertContext)
+                {
+                    int numOids;
+                    int cbData = 0;
+                    if (!Interop.crypt32.CertGetValidUsages(1, ref pCertContext, out numOids, null, ref cbData))
+                        return false;
+
+                    // -1 means the certificate is good for all usages.
+                    if (numOids == -1)
+                        return true;
+
+                    fixed (byte* pOidsPointer = new byte[cbData])
```

Comment:
In context, I feel like 'no'.  If there were no valid usages then numOids would have been -1, otherwise there would be non-zero space required for the OID extraction.